site/docs/skylark/tutorial-custom-verbs.md

```diff
@@ -0,0 +1,157 @@
+learn to recognize it.
+
+## Adapting your own rules
+
+[Starlark macros](https://docs.bazel.build/versions/master/skylark/macros.html)
+are the heart of this pattern. Macros are used like rules, but they can create
+several targets. Typically, they will create a target with the specified name
+which performs the primary build action: perhaps it builds a normal binary, a
+Docker image, or an archive of source code. In this pattern, additional targets
+are created to produce scripts performing side effects based on the output of
+the primary target, like publishing the resulting binary or updating the
+expected test output.
+
+To illustrate this, we'll wrap an imaginary rule that generates a website with
+[Sphinx](https://www.sphinx-doc.org) with a macro to create an additional
+target that allows the user to publish it when ready. Consider the following
+existing rule for generating a website with Sphinx:
+
+```python
+_sphinx_site = rule(
+     implementation = _sphinx_impl,
+     attrs = {"srcs": attr.label_list(allow_files = [".rst"])},
+)
+```
+
+Next, consider a rule like the following, which builds a script that, when run,
+publishes the generated pages:
+
+```python
+_sphinx_publisher = rule(
+    implementation = _publish_impl,
+    attrs = {
+        "site": attr.label(),
+        "_publisher": attr.label(
+            default = "//internal/sphinx:publisher",
+            executable = True,
+        ),
+    },
+    executable = True,
+)
+```
+
+Finally, we define the following macro to create targets for both of the above
+rules together:
+
+```python
+def sphinx_site(name, srcs = [], **kwargs):
+    # This creates the primary target, propducing the Sphinx-generated HTML.
+    _sphinx_site(name = name, srcs = srcs, **kwargs)
+    # This creates the secondary target, which produces a script for publishing
+    # the site generated above.
+    _sphinx_publisher(name = "%s.publish" % name, site = name, **kwargs)
+```
+
+In our `BUILD` files, we use the macro as though it just creates the primary target:
+
+```python
+sphinx_site(
+    name = "docs",
+    srcs = ["index.md", "providers.md"],
+)
+```
+
+In this example, a "docs" target is created, just as though the macro were a
+standard, single Bazel rule. When built, the rule generates some configuration
+and runs Sphinx to produce an HTML site, ready for manual inspection. However,
+an additional "docs.publish" target is also created, which builds a script for
+publishing the site. Once we check the output of the primary target, we can use
+`bazel run :docs.publish` to publish it for public consumption, just like an
+imaginary `bazel publish` command.
+
+
+It's not immediately obvious what the implementation of the `_sphinx_publisher`
+rule might look like. Often, actions like this write a _launcher_ shell script.
+This method typically involves using
+[`ctx.actions.expand_template`](https://docs.bazel.build/versions/0.28.0/skylark/lib/actions.html#expand_template)
```

Comment:
It looks like having an existing rule for this would be useful. I've filed https://github.com/bazelbuild/bazel-skylib/issues/191 for this.